# ASSOCIATION RULES

Oleh : Fitria Nur Aida

Data Source : https://www.kaggle.com/datasets/rupakroy/online-retail

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#import dataset
data=pd.read_csv("/content/drive/MyDrive/UNNES/Dataset Practice/Dataset 2/Online_Retail.csv")

In [ ]:
data.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [ ]:
data.shape

(541909, 8)

## PreProcessing

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [ ]:
#Drop Duplicate
data=data.drop_duplicates()

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 536641 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    536641 non-null  object 
 1   StockCode    536641 non-null  object 
 2   Description  535187 non-null  object 
 3   Quantity     536641 non-null  int64  
 4   InvoiceDate  536641 non-null  object 
 5   UnitPrice    536641 non-null  float64
 6   CustomerID   401604 non-null  float64
 7   Country      536641 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 36.8+ MB


In [ ]:
# Deteksi Missing Value
pd.DataFrame(np.sum(data.isnull()),columns=["Jumlah Missing"])

/usr/local/lib/python3.12/dist-packages/numpy/_core/fromnumeric.py:84: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)
  return reduction(axis=axis, out=out, **passkwargs)


Jumlah Missing
InvoiceNo                 0
StockCode                 0
Description            1454
Quantity                  0
InvoiceDate               0
UnitPrice                 0
CustomerID           135037
Country                   0

> Missing value dalam variabel Description merupakan data produk yang tidak diketahui atau tidak tercatat nama produknya dalam suatu transaksi pembelian. Oleh karena itu dilakukan penghapusan (drop) pada data missing value tersebut.

In [ ]:
# Penanganan Missing Value dengan Menghapus data missing
new=data.dropna(subset=["Description"])

In [ ]:
pd.DataFrame(np.sum(new.isnull()),columns=["Jumlah Missing"])

/usr/local/lib/python3.12/dist-packages/numpy/_core/fromnumeric.py:84: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)
  return reduction(axis=axis, out=out, **passkwargs)


Jumlah Missing
InvoiceNo                 0
StockCode                 0
Description               0
Quantity                  0
InvoiceDate               0
UnitPrice                 0
CustomerID           133583
Country                   0

> Variabel InvoiceNo terdapat data yang mengandung huruf C dan pada variabel Quantity terdapat nilai negatif yang menandakan adanya pembatalan transaksi, oleh karena itu dilakukan penghapusan data yang mengandung huruf C pada variabel InvoiceNo dan nilai negatif pada variabel Quantity.

In [ ]:
#drop InvoiceNo yang mengandung "C"
new=new[~new['InvoiceNo'].str.contains('C')]

In [ ]:
new.describe()

Quantity      UnitPrice     CustomerID
count  525936.000000  525936.000000  392732.000000
mean       10.365655       3.872616   15287.734822
std       160.075723      42.021233    1713.567773
min     -9600.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13955.000000
50%         3.000000       2.080000   15150.000000
75%        11.000000       4.130000   16791.000000
max     80995.000000   13541.330000   18287.000000

In [ ]:
#cek InvoiceNo yang mengandung "Quantity" < 0
new[new['Quantity']<0]

InvoiceNo StockCode Description  Quantity          InvoiceDate  \
7313      537032     21275           ?       -30  2010-12-03 16:50:00   
13217     537425    84968F       check       -20  2010-12-06 15:35:00   
13218     537426    84968E       check       -35  2010-12-06 15:36:00   
13264     537432    35833G     damages       -43  2010-12-06 16:10:00   
21338     538072     22423      faulty       -13  2010-12-09 14:10:00   
...          ...       ...         ...       ...                  ...   
535333    581210     23395       check       -26  2011-12-07 18:36:00   
535335    581212     22578        lost     -1050  2011-12-07 18:38:00   
535336    581213     22576       check       -30  2011-12-07 18:38:00   
536908    581226     23090     missing      -338  2011-12-08 09:56:00   
538919    581422     23169     smashed      -235  2011-12-08 15:24:00   

        UnitPrice  CustomerID         Country  
7313          0.0         NaN  United Kingdom  
13217         0.0         NaN  United Kingdom  
13218         0.0         NaN  United Kingdom  
13264         0.0         NaN  United Kingdom  
21338         0.0         NaN  United Kingdom  
...           ...         ...             ...  
535333        0.0         NaN  United Kingdom  
535335        0.0         NaN  United Kingdom  
535336        0.0         NaN  United Kingdom  
536908        0.0         NaN  United Kingdom  
538919        0.0         NaN  United Kingdom  

[474 rows x 8 columns]

In [ ]:
#cek Description yang mengandung "Quantity" < 0
display(new[new['Quantity'] < 0]['Description'].value_counts())

Description
check                    120
damages                   45
damaged                   42
?                         41
sold as set on dotcom     20
                        ... 
lost??                     1
wet                        1
wet boxes                  1
????damages????            1
lost                       1
Name: count, Length: 138, dtype: int64

In [ ]:
#drop InvoiceNo yang mengandung "Quantity" < 0
new=new[~new['Quantity']<0]

In [ ]:
new.shape

(525462, 8)

In [ ]:
new.info()

<class 'pandas.core.frame.DataFrame'>
Index: 525462 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    525462 non-null  object 
 1   StockCode    525462 non-null  object 
 2   Description  525462 non-null  object 
 3   Quantity     525462 non-null  int64  
 4   InvoiceDate  525462 non-null  object 
 5   UnitPrice    525462 non-null  float64
 6   CustomerID   392732 non-null  float64
 7   Country      525462 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 36.1+ MB


In [ ]:
# Cek Missing Value pada CustomerID
new[new['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
1447      536544     21790               VINTAGE SNAP CARDS         9   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

                InvoiceDate  UnitPrice  CustomerID         Country  
1443    2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1444    2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1445    2010-12-01 14:32:00       0.85         NaN  United Kingdom  
1446    2010-12-01 14:32:00       1.66         NaN  United Kingdom  
1447    2010-12-01 14:32:00       1.66         NaN  United Kingdom  
...                     ...        ...         ...             ...  
541536  2011-12-09 10:26:00       4.13         NaN  United Kingdom  
541537  2011-12-09 10:26:00       4.13         NaN  United Kingdom  
541538  2011-12-09 10:26:00       4.96         NaN  United Kingdom  
541539  2011-12-09 10:26:00      10.79         NaN  United Kingdom  
541540  2011-12-09 10:26:00    1714.17         NaN  United Kingdom  

[132730 rows x 8 columns]

> Missing value pada variabel CustomerID merupakan data identitas pelanggan yang tidak diketahui ID pelanggannya. ID pelanggan yang tidak diketahui diasumsikan tidak mempengaruhi data transaksi pada variabel InvoiceNo dan Description sehingga missing value tersebut tidak diatasi.

## Set Transactions Data

In [ ]:
#Set Transactions
group_InvoiceNo=new.groupby('InvoiceNo')
transactions=[]
for name,group in group_InvoiceNo:
    transactions.append(list(group["Description"].map(str)))

## Market Basket Analysis dengan Association Rules Apriori

In [ ]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=d836311a6e6c9be6c5cc50a4f78d7842f68ed275f18375e84db806c810eb1f5c
  Stored in directory: /root/.cache/pip/wheels/7f/49/e3/42c73b19a264de37129fadaa0c52f26cf50e87de08fb9804af
Successfully built apyori


In [ ]:
from apyori import apriori

In [ ]:
assrules=apriori(transactions,min_support=0.03, min_confidence=0.7, min_lift=1.1, min_length=2)
hasil=list(assrules)

In [ ]:
# Visualising the result as a readeable dataframe
final_results = pd.DataFrame(np.random.randint(low=0, high=1, size=(len(hasil), 6)), columns=['GeneralRules', 'LeftRules', 'RightRules', 'Support', 'Confidence', 'Lift'], dtype=object)
index = 0
for g, s, i in hasil:
    final_results.iloc[index] = [' _&&_ '.join(list(g)), ' _&&_ '.join(list(i[0][0])), ' _&&_ '.join(list(i[0][1])), s, i[0][2], i[0][3]]
    index = index+1
# The most significant rules
final_results1 = final_results.sort_values("Confidence", ascending=0)
final_results1

GeneralRules  \
0  PINK REGENCY TEACUP AND SAUCER _&&_ GREEN REGE...   
1  ROSES REGENCY TEACUP AND SAUCER  _&&_ GREEN RE...   

                         LeftRules                        RightRules  \
0   PINK REGENCY TEACUP AND SAUCER   GREEN REGENCY TEACUP AND SAUCER   
1  GREEN REGENCY TEACUP AND SAUCER  ROSES REGENCY TEACUP AND SAUCER    

    Support Confidence       Lift  
0  0.031436   0.826371  16.393893  
1  0.038141    0.75665  14.292598

In [ ]:
count=1
for i, d in final_results1.head(10).iterrows():
    print('Rule #'+str(count)+':')
    print(d['LeftRules'])
    print('=> '+d['RightRules'])
    print('Support: '+str(d['Support'])+' - Confidence: '+str(d['Confidence'])+' - Lift: '+str(d['Lift']))
    print('--------------------')
    count=count+1

Rule #1:
PINK REGENCY TEACUP AND SAUCER
=> GREEN REGENCY TEACUP AND SAUCER
Support: 0.031436233611442196 - Confidence: 0.8263707571801567 - Lift: 16.393893169043977
--------------------
Rule #2:
GREEN REGENCY TEACUP AND SAUCER
=> ROSES REGENCY TEACUP AND SAUCER 
Support: 0.03814064362336114 - Confidence: 0.7566502463054187 - Lift: 14.29259789831699
--------------------


> Probabilitas pembelian produk “PINK REGENCY TEACUP AND SAUCER” dan “GREEN REGENCY TEACUP AND SAUCER” secara bersamaan yaitu 0,031436 dari keseluruhan transaksi. Berdasarkan pembelian produk “PINK REGENCY TEACUP AND SAUCER”, sebesar 82,63% pelanggan juga membeli produk “GREEN REGENCY TEACUP AND SAUCER”. Nilai Lift lebih besar dari 1 sehingga pelanggan yang membeli produk “PINK REGENCY TEACUP AND SAUCER” memiliki kencerungan tinggi untuk membeli produk “GREEN REGENCY TEACUP AND SAUCER”.

> Probabilitas pelanggan membeli produk “GREEN REGENCY TEACUP AND SAUCER” dan “ROSES REGENCY TEACUP AND SAUCER” yaitu sebesar 0,038141 atau sebesar 3.8141% produk “GREEN REGENCY TEACUP AND SAUCER” dan “ROSES REGENCY TEACUP AND SAUCER” dibeli secara bersamaan dari keseluruhan transaksi. Persentase pelanggan membeli produk “ROSES REGENCY TEACUP AND SAUCER” sebesar 75,665% dari pembelian produk “GREEN REGENCY TEACUP AND SAUCER”. Sedangkan nilai Lift yang lebih besar dari 1 berarti pelanggan yang membeli produk “GREEN REGENCY TEACUP AND SAUCER” memiliki kecenderungan tinggi untuk membeli “ROSES REGENCY TEACUP AND SAUCER”.